File: C++/Brute Force/_209_MinimumSizeSubarraySum.cpp
// Problem: https://leetcode.com/problems/minimum-size-subarray-sum/

/* Solution (Brute Force):
 * 1. In this approach, we iterate through the array and keep adding the elements to the sum that are present at index j.
 * 2. If the sum is greater than or equal to the target, update the mini and break the loop.
 * 3. Here we first create all the possible subarrays by iterating using 2 pointers simultaneously.
 * 4. Return the mini.
 */

#include <bits/stdc++.h>
#include <vector>
using namespace std;

class Solution {
public:
    static int minSubArrayLen(int target, vector<int>& nums){
        int n = nums.size();
        int mini = INT_MAX;

        for(int i = 0; i < n; i++){
            int sum = 0;
            for(int j = i; j < n; j++){
                sum += nums[j];
                if(sum >= target){
                    mini = min(mini, j - i + 1);
                    break;
                }
            }
        }

        return mini == INT_MAX ? 0 : mini;
    }
};

int main() {
    vector<int> nums = {2,3,1,2,4,3};
    int target = 7;

    cout << Solution::minSubArrayLen(target, nums) << endl;
    return 0;
}

// Time Complexity: O(n^2)
// Space Complexity: O(1)

File: JAVA/_209_MinimumSizeSubarraySum.java
// Problem: https://leetcode.com/problems/minimum-size-subarray-sum/

/* Solution (Optimal Solution): using Sliding window technique
 * 1. In this approach, we keep only the desired subarray between the sliding window.
 * 2. Iterate through the array and keep adding the elements to the sum that are present at index R.
 * 3. If the sum is greater than or equal to the target, update the mini and remove the value at l index. Do this until the sum is not greater than or equal to the target.
 * 4. After each iteration, move only the pointer R to the right till it reaches the end of the array.
 * 5. Return the mini.
*/


public class _209_MinimumSizeSubarraySum {
    private static int minSubArrayLen(int target, int[] nums){
        int mini = Integer.MAX_VALUE;
        int n = nums.length;
        int l = 0; 
        int sum = 0;

        for(int r = 0; r < n; r++){
            sum += nums[r];
            while(sum >= target){
                mini = Math.min(mini, r - l + 1);
                sum -= nums[l];
                l++;
            }
        }
        return mini == Integer.MAX_VALUE ? 0 : mini;
    }
    public static void main(String[] args){
        int[] nums = {1, 1, 1, 1, 1, 1, 1};
        int target = 11;

        System.out.println(minSubArrayLen(target, nums));
    }
}

// Time Complexity: O(nlogn)
// Space Complexity: O(1)